{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import GenBank\n",
    "import re as re\n",
    "from pathlib import Path"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Constants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nb_of_segments = NUMBER_OF_SEGMENTS -> int\n",
    "nt = PERMISSIBLE_OF_NUCLEOTIDE -> int         \n",
    "nb_segments = {\n",
    "    '1':  QUANTITY OF NUCLEOTIDES IN SEGMENT 1 -> int ,\n",
    "    '2':  QUANTITY OF NUCLEOTIDES IN SEGMENT 2  -> int ,\n",
    "    '3':  QUANTITY OF NUCLEOTIDES IN SEGMENT 3 -> int ,\n",
    "    '4':  QUANTITY OF NUCLEOTIDES IN SEGMENT 4 -> int ,\n",
    "    '5':  QUANTITY OF NUCLEOTIDES IN SEGMENT 5 -> int ,\n",
    "    '6':  QUANTITY OF NUCLEOTIDES IN SEGMENT 6 -> int ,\n",
    "    '7':  QUANTITY OF NUCLEOTIDES IN SEGMENT 7 -> int ,\n",
    "    '8':  QUANTITY OF NUCLEOTIDES IN SEGMENT 8 -> int ,\n",
    "                }"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Functions that we used"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#in some cases, segment names consist of letters, we need to change them to numbers\n",
    "def change_letters_to_numbers(organisms: dict) -> dict:\n",
    "    for seg_seq in organisms.values():\n",
    "            not_allowed = []\n",
    "            if 'M' in seg_seq.keys() and 'L' in seg_seq.keys() and 'S' in seg_seq.keys():\n",
    "                    seg_seq['1']=seg_seq['S']\n",
    "                    seg_seq.pop('S')\n",
    "                    seg_seq['2']=seg_seq['M']\n",
    "                    seg_seq.pop('M')\n",
    "                    seg_seq['3']=seg_seq['L']\n",
    "                    seg_seq.pop('L')\n",
    "            elif 'S' in seg_seq.keys() and 'L' in seg_seq.keys() and 'M' not in seg_seq.keys():\n",
    "                    seg_seq['1']=seg_seq['S']\n",
    "                    seg_seq.pop('S')\n",
    "                    seg_seq['2']=seg_seq['L']\n",
    "                    seg_seq.pop('L')\n",
    "            not_allowed = list(filter(None, \"\".join(re.split(r'\\d', \"/\".join(seg_seq.keys()))).replace('\"',\"\").split(\"/\")))\n",
    "            \n",
    "            if len(not_allowed) != 0:\n",
    "                for wrong_name in not_allowed:\n",
    "                      seg_seq.pop(wrong_name)\n",
    "           \n",
    "    return organisms\n",
    "#we have to choose the longest segment\n",
    "\n",
    "def choose_the_proteint_coding_sequence(lib: dict) -> dict):\n",
    "    for name, content in lib.items():\n",
    "        for num, data in content.items():\n",
    "            #print(data)\n",
    "            if len(data) > 1:\n",
    "                \n",
    "                \n",
    "                good = []\n",
    "                for item in data[1]:\n",
    "                    if \"(\" in item:\n",
    "                        good.extend(item.split(\"(\")[1][:-1].split(\",\"))\n",
    "                    else:\n",
    "                        good.append(item)\n",
    "                max = 0\n",
    "                best = \"\"\n",
    "                #print(good)\n",
    "                for item in good:\n",
    "                    t = item.split(\"..\")\n",
    "                    #print(item)\n",
    "                    i = int(t[0].strip(\"<>\"))\n",
    "                    j = int(t[1].strip(\"<>\"))\n",
    "                    if j - i > max:\n",
    "                        max = j-i\n",
    "                        best = item\n",
    "                \n",
    "                data[1] = best.replace(\">\",\"\").replace(\"<\",\"\")\n",
    "                \n",
    "            else:\n",
    "                data = []\n",
    "    \n",
    "    return lib\n",
    "                \n",
    "\n",
    "#sometimes we should cut nucleotides before first start-codon and after the last stop-codon\n",
    "def cut_primers(organisms: dict,  nb_of_segments: dict) -> dict:\n",
    "\n",
    "    for dictt in organisms.values():\n",
    "        for number_of_segment, listt in dictt.items():\n",
    "\n",
    "            if number_of_segment == '1':\n",
    "                if len(listt) > 1:\n",
    "                    cut_start = listt[-1].split(\"..\")[0]\n",
    "                    #if cut_start not in [\"A\", \"T\", \"G\", \"C\", \"N\"]:\n",
    "                    cut_start = int(cut_start) - 1\n",
    "\n",
    "                    dictt[number_of_segment] = listt[0][cut_start::]\n",
    "               \n",
    "            if int(number_of_segment.replace('\"', \"\")) == nb_of_segments:\n",
    "                if len(listt) > 1:\n",
    "                    cut_end = listt[-1].split(\"..\")[-1]\n",
    "                    #if cut_end not in [\"A\", \"T\", \"G\", \"C\"]:\n",
    "                    cut_end = int(cut_end)\n",
    "                    dictt[number_of_segment] = listt[0][0:cut_end]\n",
    "\n",
    "            elif number_of_segment !='1' and number_of_segment != nb_of_segments:\n",
    "                  dictt[number_of_segment] = listt[0]\n",
    "                     \n",
    "    return organisms\n",
    "\n",
    "\n",
    "\n",
    "def cut_primers_for_all_segments(organisms: dict) -> dict:\n",
    "\n",
    "    for dictt in organisms.values():\n",
    "        for number_of_segment, listt in dictt.items():\n",
    "\n",
    "                if len(listt) > 1:\n",
    "                    cut_start = listt[-1].split(\"..\")[0]\n",
    "                    cut_start = int(cut_start) - 1\n",
    "                    cut_end = listt[-1].split(\"..\")[-1]\n",
    "                    cut_end = int(cut_end)\n",
    "                    \n",
    "                    dictt[number_of_segment] = listt[0][cut_start:cut_end]\n",
    "\n",
    "                     \n",
    "    return organisms\n",
    "\n",
    "\n",
    "#check len of segments\n",
    "def check_segments(organisms: dict, nb_segments: int, nt: int) -> tuple(dict, dict):\n",
    "    seq2 = {}\n",
    "    length_of_segments = {}\n",
    "    for name, dictt in organisms.items():\n",
    "        seq2[name] = {}  \n",
    "        length_of_segments[name] = {}\n",
    "        for k, v in nb_segments.items():  \n",
    "             if k in dictt.keys():\n",
    "                length_of_segments[name][k] = len(dictt[k]) \n",
    "                if (v - nt <= len(dictt[k]) <= v + nt):\n",
    "                             seq2[name][k]=dictt[k]\n",
    "                             \n",
    "\n",
    "    return seq2, length_of_segments                 \n",
    "\n",
    "#here we bring to the form {id:record of all segments}\n",
    "def prep_delete_degenerate_nucleotides(seq2: dict, nb_of_segments: int) -> dict:\n",
    "    prep_without_dn = {}\n",
    "    for key, values in seq2.items():\n",
    "        if len(values) == nb_of_segments:\n",
    "         prep_without_dn[key] = \"\".join(str(dict(sorted(values.items(), key=lambda x: x[0])).values())).replace(\"'\",\"\").replace('dict_values([',\"\").replace('])',\"\").replace(',',\"\").replace(\" \",\"\")\n",
    "         #prep_without_dn[key] = dict(sorted(values.items(), key=lambda x: x[0])).values()\n",
    "   \n",
    "    return prep_without_dn\n",
    "\n",
    "#sometimes we should delete {id:record} with degenerate nucleotides\n",
    "def delete_degenerate_nucleotides(prep_without_dn: dict) -> dict:\n",
    "    without_dn={}\n",
    "    dn = ['u', 'w', 'n', 's', 'm', 'k', 'r', 'y', 'b', 'd', 'h', 'v']\n",
    "    for key3, value3 in prep_without_dn.items():\n",
    "        y = 0\n",
    "        for nt in dn:\n",
    "            if nt in value3.lower():\n",
    "                y = 1\n",
    "        if y == 0:\n",
    "            without_dn[key3]=value3\n",
    "    return without_dn\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "App"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(path_to_your_genbank, \"r\") as handle:\n",
    "                organisms = {}\n",
    "                for record in GenBank.parse(handle):\n",
    "                    not_protein = []  \n",
    "                    future_id_all = [\"\"]\n",
    "                    number_of_segment = 'wrong'\n",
    "                    segments = [] \n",
    "                    strings = []\n",
    "\n",
    "                    for feature in record.features:\n",
    "                            fut_id_var1 = \"\"\n",
    "                            fut_id_var2 = \"\"\n",
    "                            fut_id_var3 = \"\"\n",
    "                            if feature.key == \"CDS\":\n",
    "                                 not_protein.append(feature.location)\n",
    "                                 \n",
    "                            for qualifier in feature.qualifiers:  \n",
    "                        \n",
    "                                if qualifier.key == \"/segment=\":  \n",
    "                                    number_of_segment = qualifier.value.replace('\"',\"\")\n",
    "                                    segments.append(number_of_segment)\n",
    "                                if qualifier.key == \"/strain=\":  \n",
    "                                     fut_id_var1 = qualifier.value\n",
    "                                     future_id_all.append(fut_id_var1)\n",
    "                                if qualifier.key ==  \"/isolate=\":\n",
    "                                     fut_id_var2 = qualifier.value\n",
    "                                     future_id_all.append(fut_id_var2)\n",
    "                                #may be useful for some case\n",
    "                                if qualifier.key == \"/organism=\":\n",
    "                                     fut_id_var3 = qualifier.value\n",
    "                                     future_id_all.append(fut_id_var3) \n",
    "                                fut_id = max(future_id_all, key = len)\n",
    "                            \n",
    "                            if fut_id not in organisms.keys():\n",
    "                                 organisms[fut_id] = {}\n",
    "                    organisms[fut_id][number_of_segment] = [record.sequence]\n",
    "                    if not_protein:\n",
    "                        organisms[fut_id][number_of_segment].append(not_protein)\n",
    " \n",
    "            organisms = change_letters_to_numbers(organisms)\n",
    "            organisms = choose_the_proteint_coding_sequence(organisms)\n",
    "            #if you want only protein-coding-sequence:\n",
    "                organisms = cut_primers_for_all_segments(organisms)\n",
    "            #else (cut primers from first segment before start-codon and after the last stop-codon):\n",
    "                organisms = cut_primers(organisms, nb_of_segments)\n",
    "            organisms_right_len, length_of_segments = check_segments(organisms, nb_segments, nt)\n",
    "            prep_without_dn = prep_delete_degenerate_nucleotides(organisms_right_len, nb_of_segments)\n",
    "            without_dn = delete_degenerate_nucleotides(prep_without_dn)\n",
    "\n",
    "            #create a file with mergening genome of virus\n",
    "            with open(\"all_segments.fasta\", \"w\") as f:\n",
    "                for key1, item in  without_dn.items():\n",
    "                    strings.append(\">{} \\n{}\\n\".format(key1, item))\n",
    "                result1 = \"\".join(strings)\n",
    "                f.write(result1)"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
